File: catkin_ws/src/odometry/src/odometry_publisher.py
#!/usr/bin/env python3

import rospy
import serial
import math
from sensor_msgs.msg import Imu
from nav_msgs.msg import Odometry

# Constants
WHEEL_BASE = 0.54  # Distance between left and right wheels (meters)

# Global Variables
prev_right_distance = 0.0
prev_left_distance = 0.0
x, y, th = 0.0, 0.0, 0.0  # Position & Orientation

# Initialize Serial Connection
try:
    ser = serial.Serial('/dev/ttyACM0', 115200, timeout=1)
    rospy.loginfo("Serial port /dev/ttyACM0 opened successfully.")
except serial.SerialException:
    rospy.logerr("Failed to open serial port! Check connection.")
    exit()

def publish_imu_data(vx, vth):
    """Publishes velocity as /imu/data (even though it's from wheel encoders)."""
    imu_msg = Imu()
    imu_msg.header.stamp = rospy.Time.now()
    imu_msg.header.frame_id = "imu_link"

    # Assign linear and angular velocity
    imu_msg.linear_acceleration.x = vx  # Forward velocity (m/s)
    imu_msg.angular_velocity.z = vth    # Rotational velocity (rad/s)

    imu_pub.publish(imu_msg)
    rospy.loginfo("Publishing /imu/data -> Linear: %.4f m/s | Angular: %.4f rad/s", vx, vth)

def publish_odometry(vx, vth, dt):
    """Computes and publishes odometry to /odom."""
    global x, y, th, prev_right_distance, prev_left_distance

    # Compute position update
    delta_s = vx * dt
    delta_th = vth * dt

    x += delta_s * math.cos(th + delta_th / 2.0)
    y += delta_s * math.sin(th + delta_th / 2.0)
    th += delta_th
    th = math.atan2(math.sin(th), math.cos(th))  # Normalize theta

    # Create Odometry Message
    odom_msg = Odometry()
    odom_msg.header.stamp = rospy.Time.now()
    odom_msg.header.frame_id = "odom"
    odom_msg.child_frame_id = "base_link"

    odom_msg.pose.pose.position.x = x
    odom_msg.pose.pose.position.y = y
    odom_msg.pose.pose.orientation.z = math.sin(th / 2.0)
    odom_msg.pose.pose.orientation.w = math.cos(th / 2.0)

    odom_msg.twist.twist.linear.x = vx
    odom_msg.twist.twist.angular.z = vth

    odom_pub.publish(odom_msg)
    rospy.loginfo("Publishing /odom -> x: %.4f, y: %.4f, θ: %.4f", x, y, th)

def main():
    global prev_right_distance, prev_left_distance

    rospy.init_node("odometry_publisher")
    global imu_pub, odom_pub
    imu_pub = rospy.Publisher("/imu/data", Imu, queue_size=10)
    odom_pub = rospy.Publisher("/odom", Odometry, queue_size=10)

    rate = rospy.Rate(10)  # 10 Hz loop
    last_time = rospy.Time.now()

    while not rospy.is_shutdown():
        current_time = rospy.Time.now()
        dt = (current_time - last_time).to_sec()
        last_time = current_time

        # Read Serial Data from Encoders
        if ser.in_waiting:
            serial_data = ser.readline().decode().strip()
            rospy.loginfo("Raw Serial Data: %s", serial_data)

            try:
                parts = serial_data.split(',')
                right_distance = float(parts[0].split(':')[1])
                left_distance = float(parts[1].split(':')[1])

                # Compute per-loop displacement (ΔD)
                delta_s_right = right_distance - prev_right_distance
                delta_s_left = left_distance - prev_left_distance

                # Update previous displacement values
                prev_right_distance = right_distance
                prev_left_distance = left_distance

                # Compute forward velocity and angular velocity
                vx = (delta_s_right + delta_s_left) / (2.0 * dt)
                vth = (delta_s_right - delta_s_left) / (WHEEL_BASE * dt)

                # Publish IMU Data (Wheel-based velocity)
                publish_imu_data(vx, vth)

                # Publish Odometry
                publish_odometry(vx, vth, dt)

            except Exception as e:
                rospy.logwarn("Malformed Serial Data: %s | Error: %s", serial_data, str(e))

        rate.sleep()

if __name__ == "__main__":
    try:
        main()
    except rospy.ROSInterruptException:
        pass

File: catkin_ws/src/odometry_publisher/src/odometry_publisher.cpp
#include <ros/ros.h>
#include <tf/transform_broadcaster.h>
#include <nav_msgs/Odometry.h>
#include <serial/serial.h> // Serial communication for Jetson-Arduino
#include <cmath>

// Global Variables
double x = 0.0, y = 0.0, th = 0.0; // Position & orientation
double prev_right_distance = 0.0, prev_left_distance = 0.0; // Store previous encoder readings

const double WHEEL_BASE = 0.54; // Distance between wheels (meters)
serial::Serial ser; // Serial communication object

int main(int argc, char** argv) {
    ros::init(argc, argv, "odometry_publisher");
    ros::NodeHandle n;

    ros::Publisher odom_pub = n.advertise<nav_msgs::Odometry>("odom", 50);
    tf::TransformBroadcaster odom_broadcaster;

    // Open Serial Port
    try {
        ser.setPort("/dev/ttyACM0"); // Adjust as needed
        ser.setBaudrate(115200);
        serial::Timeout timeout = serial::Timeout::simpleTimeout(1000);
        ser.setTimeout(timeout);
        ser.open();
    } catch (serial::IOException& e) {
        ROS_ERROR("Unable to open serial port!");
        return -1;
    }

    ros::Time last_time = ros::Time::now();
    ros::Rate r(10.0); // 10 Hz loop

    while (ros::ok()) {
        ros::spinOnce();
        ros::Time current_time = ros::Time::now();
        double dt = (current_time - last_time).toSec();
        last_time = current_time;

        // Read total displacement from encoders
        double right_distance = 0.0, left_distance = 0.0;
        if (ser.available()) {
            std::string serial_data = ser.readline();
            if (sscanf(serial_data.c_str(), "Right Displacement:%lf,Left Displacement:%lf", 
                       &right_distance, &left_distance) == 2) {
                // Successfully read total displacement values
            }
        }

        // 🔹 Compute per-loop displacement (ΔD) by subtracting previous readings
        double delta_s_right = right_distance - prev_right_distance;
        double delta_s_left = left_distance - prev_left_distance;

        // Update previous displacement values for next iteration
        prev_right_distance = right_distance;
        prev_left_distance = left_distance;

        // 🔹 Compute forward displacement and rotation
        double delta_s = (delta_s_right + delta_s_left) / 2.0;
        double delta_th = (delta_s_right - delta_s_left) / WHEEL_BASE;

        // 🔹 Update robot position
        x += delta_s * cos(th + delta_th / 2.0);
        y += delta_s * sin(th + delta_th / 2.0);
        th += delta_th;

        // Normalize theta to [-pi, pi]
        th = atan2(sin(th), cos(th));

        geometry_msgs::Quaternion odom_quat = tf::createQuaternionMsgFromYaw(th);

        // 🔹 Compute linear and angular velocity
        double vx = delta_s / dt;
        double vth = delta_th / dt;

        // Publish odometry message
        nav_msgs::Odometry odom;
        odom.header.stamp = current_time;
        odom.header.frame_id = "odom";
        odom.pose.pose.position.x = x;
        odom.pose.pose.position.y = y;
        odom.pose.pose.position.z = 0.0;
        odom.pose.pose.orientation = odom_quat;
        odom.child_frame_id = "base_link";
        odom.twist.twist.linear.x = vx;  // Correct velocity calculation
        odom.twist.twist.angular.z = vth;
        odom_pub.publish(odom);

        r.sleep();
    }
}

